forecasttools/trajectories_to_quantiles.py

```diff
@@ -0,0 +1,57 @@
+import numpy as np
+import polars as pl
+
+
+def trajectories_to_quantiles(
+    trajectories: pl.DataFrame,
+    quantiles: list[int] = None,
+    timepoint_cols: list[str] = ["timepoint"],
+    value_col: str = "value",
+    quantile_value_name: str = "quantile_value",
+    quantile_level_name: str = "quantile_level",
+    id_cols: list[str] = None,
+) -> pl.DataFrame:
+    # set default quantiles
+    if quantiles is None:
+        quantiles = (
+            [0.01, 0.025] + list(np.arange(0.05, 1.0, 0.05)) + [0.975, 0.99]
+        )
+    # group trajectories based on timepoint_cols and id_cols
+    group_cols = (
+        timepoint_cols if id_cols is None else timepoint_cols + id_cols
+    )
+
+    # compute quantiles
+    def compute_quantiles(values):
+        quantile_values = np.quantile(values, quantiles)
+        return quantile_values
+
+    # get quantiles across epiweek for forecast
+    quantile_results = trajectories.group_by(group_cols).agg(
+        [
+            pl.col(value_col)
+            .map_elements(
+                lambda group_values: compute_quantiles(group_values),
+                return_dtype=pl.List(pl.Float64),
+            )
+            .alias("quantile_values")
+        ]
+    )
+    # resultant quantiles into individual rows
+    quant_df = quantile_results.explode("quantile_values")
+    # aligning rows with quantile levels
+    quant_df = quant_df.with_columns(
+        [
+            pl.Series(quantiles * (len(quant_df) // len(quantiles))).alias(
+                "quantile_levels"
+            )
+        ]
+    )
```

Comment:
This seems fragile. Better to keep the quantile id and quantile paired together at the time of creation. In R:

``` r
library(tidyverse)

quantiles <- seq(0.1, 0.9, 0.1)
tibble(my_dates = seq(today() - 7, today(), by = "1 day")) |> 
  mutate(x = map(1:n(), ~rnorm(1000))) |> 
  mutate(quant = map(x, \(x) quantile(x, probs = quantiles))) |> 
  unnest_longer(quant) |> 
  select(-x)
#> # A tibble: 72 × 3
#>    my_dates     quant quant_id
#>    <date>       <dbl> <chr>   
#>  1 2024-09-30 -1.30   10%     
#>  2 2024-09-30 -0.868  20%     
#>  3 2024-09-30 -0.521  30%     
#>  4 2024-09-30 -0.232  40%     
#>  5 2024-09-30  0.0305 50%     
#>  6 2024-09-30  0.289  60%     
#>  7 2024-09-30  0.530  70%     
#>  8 2024-09-30  0.885  80%     
#>  9 2024-09-30  1.26   90%     
#> 10 2024-10-01 -1.21   10%     
#> # ℹ 62 more rows
```

<sup>Created on 2024-10-07 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
I agree it is fragile and there is likely some way in `polars` to do this but I just wrote what worked in my instance. Was expecting this to get addressed via this PR.

Reply:
For your reference @damonbayer [in R I did it with `dplyr::reframe`](https://github.com/cdcent/cfa-forecasttools/blob/main/R/trajectories_to_quantiles.R)

Reply:
In polars maybe an `agg()` that yields one column per quantile level followed by an `unpivot()`? Or as @damonbayer creating a single struct valued column with all the quantiles and then `unnest()`-ing.

Reply:
@dylanhmorris Will investigate these.

Reply:
e.g.
```Python
import polars as pl
import numpy as np

df = pl.DataFrame({"value": np.random.random(size=100),
                                  "label": np.random.choice(["a", "b", "c"], size=100)})
quantile_levels = [0.1, 0.2, 0.4, 0.5, 0.6, 0.8, 0.9]

df.group_by("label").agg(
    [pl.col("value").quantile(x).alias(f"{x}") for x in quantile_levels]
).unpivot(index="label", variable_name="quantile_level")
```

Reply:
A note here because I accidentally resolved this without originally addressing it. All Dylan's example needed was switching interpolation from "nearest" to "midpoint" (argument of `quantile()`) and adding "value_name" to the `unpivot`. 

```python
example_df = pl.from_dict(
    {
        "epiweek": [1, 1, 1, 2, 2, 2],
        "epiyear": [2023] * 6,
        ".draw": [1, 1, 1, 2, 2, 2],
        "location": ["TX"] * 6,
        "weekly_hosp": [124, 45, 234, 23, 60, 100],
    }
)
value_col = "weekly_hosp"
group_cols = ["epiweek", "epiyear", "location"]
quantiles = [0.25, 0.50, 0.75]
# method 01
quant_df_01 = (
    example_df.group_by(group_cols)
    .agg(
        [
            pl.col(value_col)
            .map_elements(
                lambda vals: np.quantile(vals, quantiles),
                return_dtype=pl.List(pl.Float64),
            )
            .alias("quantile_values")
        ]
    )
    .explode("quantile_values")
)
quant_df_01 = quant_df_01.with_columns(
    [
        pl.Series(quantiles * (len(quant_df_01) // len(quantiles))).alias(
            "quantile_level"
        )
    ]
).sort(["epiweek", "quantile_level"])
# method 02
quant_df_02 = (
    example_df.group_by(group_cols)
    .agg(
        [
            pl.col(value_col)
            .quantile(x, interpolation="midpoint")
            .alias(f"{x}")
            for x in quantiles
        ]
    )
    .unpivot(
        index=group_cols,
        variable_name="quantile_level",
        value_name="quantile_values",
    )
    .with_columns(pl.col("quantile_level").cast(pl.Float64))
    .sort(["epiweek", "quantile_level"])
)
print(quant_df_02)
```

<img width="678" alt="quant_df_01" src="https://github.com/user-attachments/assets/0ab740ab-6d02-4601-9f4d-80ece9be6842">

<img width="649" alt="quant_df_02_mid" src="https://github.com/user-attachments/assets/095bb7e9-332e-4689-97e7-3f967352f155">